Improve ShapeDecoration performance.

bernaferrari opened the pull request:
Discussion: https://github.com/flutter/flutter/pull/108052

I believe that PR became too big and it is doing too many things at once. This PR is not breaking but might require a g3fix. So I'm splitting it. Easier to fix and easier to track the performance gains.

cc @Hixie. Same code you already reviewed (with an additional comment in `box_border.dart`).

Edit: I could make this change (really small optimization, only render border if width != 0) yet the code looks horrible. So I'm not making it, yet.
<img width="822" alt="image" src="https://user-images.githubusercontent.com/351125/182041165-98f8c922-3018-42e3-b58a-0144825bb16b.png">

bernaferrari committed:
Improve ShapeDecoration performance.

bernaferrari commented:
I am so confused. The golden was added a few days ago andthere doesn't seem to be any visible difference. I guess we can ignore it?

Hixie commented:
Yeah it is indeed a change in antialiasing. Maybe drawCircle and drawPath differ slightly?

I've marked it as OK.

Hixie commented:
> Edit: I could make this change (really small optimization, only render border if width != 0) yet the code looks horrible. So I'm not making it, yet.

Yeah I would not make that change either.

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -533,6 +533,76 @@ abstract class ShapeBorder {
+  /// true. Subclasses should only override [paintInterior] if doing so enables
+  /// faster
+  /// rendering than is possible with [Canvas.drawPath] (so, in particular,
+  /// subclasses should not call [Canvas.drawPath] in [paintInterior]).
```
trivial nit: this paragraph needs reflowing

bernaferrari replied:
Sharp eyes. Fixed!

bernaferrari committed:
Fix paragraph.

Hixie commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -180,6 +180,17 @@ abstract class BoxBorder extends ShapeBorder {
 
+  @override
+  void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+    // If `ShapeDecoration(shape: Border.all())`, ShapeDecoration doesn't
```
If -> For?

Maybe clarify this further, explaining that this class only ever paints a rounded border if `paint` is given one and that this is an extension to the superclass' API.

bernaferrari replied:
Thanks! I re-wrote it:
```
    // For `ShapeDecoration(shape: Border.all())`, a rectangle with sharp edges
    // is always painted. There is no BorderRadius parameter for
    // ShapeDecoration or Border, only for BoxDecoration, which doesn't
    // call this method.
```

Hixie commented on packages/flutter/lib/src/painting/shape_decoration.dart:
```diff
@@ -336,19 +336,28 @@ class _ShapeDecorationPainter extends BoxPainter {
+              _decoration.shape.getOuterPath(rect.shift(shadow.offset).inflate(shadow.spreadRadius), textDirection: textDirection),
+              shadow.toPaint(),
+            );
+          }),
```
i guess technically we're computing the paints more than we need to (we only need to compute them the first time, not when rect changes). Maybe add a TODO? Or not, your call. It's not a big deal.

bernaferrari replied:
I need test, but what if rotation happens? Is it the same shadow?

bernaferrari replied:
I'll try later (possibly tomorrow) but if get this working, optimization can be really big.

Hixie replied:
well the _paint_ definitely isn't affected by the rect (it's computed by `shadow.toPaint()`, which doesn't get the rect). But the question is whether it's more expensive to compute the paint, or more expensive to handle the overhead of storing the paint in some other data structure in parallel with the rect bounds / paths. IIRC you had it the other way before and it was slower?

bernaferrari replied:
I believe it was slower because of the additional loop, not exactly the additional structure. In this case, everything is going to be mostly similar. But I'll test, let's see.

bernaferrari replied:
One thing we could also do is not-cache the shadow path for primitive types (rectangle/stadium/circle), so later we can call `paintInterior` for the shadow. Your question will be answered with this. If performance is better, it is `drawPath`, if performance is worse, it is `cachePath` that takes longer.

bernaferrari replied:
I have a proposal.

What if we had a `mixin PrimitiveShape { paintInterior(...) } ` that is applied on RoundedRectangleBorder / CircleBorder / Stadium Border, then we do `if (shape is PrimitiveShape) { (shape as PrimitiveShape).paintInterior(...) }`?

We wouldn't need `preferPaintInterior`. And... maybe, one day, we could delimit the `BorderSide.only`/`MultipleBorderSide` (I still need to convince you) to only work on shapes that have that mixin. Is this too crazy?

bernaferrari replied:
Update: your proposal already happens. `interiorPaint` is only set when not null. Gradient, image and shadow depend on the rect. So it is already super optimized. 

<img width="830" alt="image" src="https://user-images.githubusercontent.com/351125/182449700-e087bb52-cad7-41e2-9f34-74af4b419823.png">

bernaferrari replied:
Found what you said. Now I understand the issue. Testing a few things.

bernaferrari replied:
After debugging today: negligible difference, I'm reverting for the old way. Not changing what already works fine. My sample only uses a single color, if paint is complex enough, there might be differences.

Again, you are correct. Thanks!

Hixie reviewed (approved):
![LGTM](http://junkyard.damowmow.com/578)

goderbauer commented on packages/flutter/lib/src/painting/borders.dart:
````diff
@@ -533,6 +533,76 @@ abstract class ShapeBorder {
+  /// }
+  /// ```
+  ///
+  /// When a shape can only be drawn using path, [preferPaintInterior] should
````
should -> must

goderbauer reviewed (approved):
LGTM

goderbauer commented:
@bernaferrari What is the status of this PR? It looks like the three last comments above may still be unresolved?

bernaferrari commented:
I traveled for the past few days, so couldn't find time to run some additional benchmarks to see if a variable was better or worse than a loop. I'll probably take a look tomorrow.

I also made a proposal to use a mixin instead of the hasPaintInside getter, but didn't get an answer.

Regardless, everything should be resolved on the next few days.

bernaferrari committed:
Fix goderbauer review.

bernaferrari committed:
Reword Border paintInterior.

bernaferrari committed:
Revert optimization.

bernaferrari committed:
Fix line.

bernaferrari commented:
Sorry for the delay, here we go!! Fixed what needed to be fixed.

I rewrote this, if it is ok, feel free to add auto-submit. If it is not, let me know how I can improve. That's the last thing.
https://github.com/flutter/flutter/blob/d742c5a2e66e3fe18ce14dc5b1d0c8bb02c355ad/packages/flutter/lib/src/painting/box_border.dart#L185-L188

bernaferrari committed:
Fix documentation with `abstract`.

bernaferrari commented:
I also made the snippet `abstract` so it compiles without implementing everything else, now that the snippet rules changed:
<img width="355" alt="image" src="https://user-images.githubusercontent.com/351125/185032711-710833f2-7991-4986-a378-aac6a404cf54.png">

bernaferrari committed:
Fix analyzer.

Hixie commented:
> I also made the snippet `abstract` so it compiles without implementing everything else

You don't need the `abstract` if you have the `// ...` bit (it adds an `ignore` for implementing abstract methods).

bernaferrari committed:
Remove abstract

bernaferrari commented:
Thanks, I didn't know that.

Hixie commented:
> Thanks, I didn't know that.

Don't worry, you have no real way to have found out. It's documented in the bottom of a locked filing cabinet stuck in a disused lavatory with a sign on the door saying "Beware of the Leopard": https://github.com/flutter/flutter/blob/master/dev/bots/analyze_snippet_code.dart#L51

bernaferrari committed:
`BorderRadius` -> `borderRadius` in the comment
